plugin/completion.py

```diff
@@ -260,7 +255,10 @@ def do_request(self, prefix: str, locations: 'List[int]') -> None:
         if not client:
             return
 
-        if settings.complete_all_chars or self.is_after_trigger_character(locations[0]):
+        prev_point = locations[0] - 1 if locations[0] - 1 >= 0 else 0
+        prev_char = view.substr(prev_point)
+
+        if prev_char in self.trigger_chars or prev_char not in self.ignored_chars:
```

Comment:
Perhaps we can keep the setting as it is, but change its behavior so that it only triggers for word characters. We already have a function for that: https://github.com/sublimelsp/LSP/blob/96e28970c48680d00141b88631feedcb9c141cf2/plugin/core/documents.py#L46-L51

```suggestion
        if prev_char in self.trigger_chars or (settings.complete_all_chars and position_is_word(self.view, prev_point)):
```

Reply:
I tried using `position_is_word` :) 
Here is what the problem is with that approach.

If I use:
```python
def do_request(self, prefix: str, locations: 'List[int]') -> None:
        ...
        if prev_char in self.trigger_chars or(settings.complete_all_chars and position_is_word(self.view, prev_point)):
```
and I have a file empty python file. (the cursor is `|`)
```python
| 
```
And I manually trigger the autocomplete.
I will get no completions.
But I expect to see completions.

Because the previous cursor character is not a trigger char
and the character under cursor is not a word. 

With this PR, as it is now, 
you will get completions suggestion for, import, assert, class, def, ... 

Second example:
Lets say we have a python file:

```
# there is one space between cursor and 1
if 2 > 1 |
```
And you manually trigger the autocomplete.
You will get will get no completions.
Because the previous cursor character is not a trigger char
and the character under cursor is not a word. 

With this PR as it is now, you will get completions suggestion for, and, is, in, not.

The problem is even if we have completions at some position, 
and that position is not in the trigger chars and not a word,
when we trigger the autocomplete manually, 
we will get no results.

The solution to that problem is that we need a flag, `is_autocomplte_manually` in the `do_request` function.
 And if it is manually triggered, we will send the completion request to the server,
regardless if the character under cursor is a:
- trigger char 
- or a word 
- or neither of both.

That means no matter the case, if the autocomplete was triggered manually, we should send a completion request.

But I don't know how to determine if autocomplete was trigger manually or not. 
Or to be more precise, 
I don't know how determine if the `on_query_command` was called by the user pressing `crtl+space` or by the plugin like in the on_modified() defined in `CompletionHandler` class.
```
    def on_modified(self) -> None:
        ...
        if self.committing:
            self.committing = False
            self.on_completion_inserted()
        else:
            if self.view.is_auto_complete_visible():
                if self.response_incomplete:
                    # debug('incomplete, triggering new completions')
                    self.view.run_command("hide_auto_complete") # this here will hide the autocomplete
                    sublime.set_timeout(self.run_auto_complete, 0) # this will trigger on_query_completion
```

So instead of asking what characters should trigger the autocomplete,
I ask the question what characters should not trigger the autocomplete.
That is the reason why I did it this way.

But If you know a way of how to determine if the autocomplete was triggered manually,
let me know :)

Reply:
Good point about the manual autocomplete. Still, we should strive for no extra settings. Every view has a setting called `word_separators`, perhaps we can use that?

```
>>> view.settings().get("word_separators")
'./\\()"\'-:,.;<>~!@#$%^&*|+=[]{}`~?'
```

Reply:
I removed the `auto_complete_ignored_triggers` setting and brought back `complete_all_chars` (although I have nothing against removing that setting too).

I also tried to simplify the completion file,
there was a lot of logic there.
Hope people will understand it easily now.

I made sure that all functionality is still there, but this need to be tested. I tested it with, LSP-typescript, LSP-intelephense(with this server I tested the is_incomplete flag), pyls, LSP-json...

Everything seems ok, 
but I would not merge this, until I test this for a week of real usage.

Reply:
There are already a bunch of completion tests. Run "UnitTesting: Test Current File" from the command palette when you've opened tests/test_completion.py to check and expand your work.

Reply:
thanks for the tip